### PR TITLE
[SPARK-34541][CORE] Fixed an issue where data could not be cleaned up when unregisterShuffle.

### PR DESCRIPTION
**Pull request by yikf**
### What changes were proposed in this pull request?
Fixed an issue where data could not be cleaned up when unregisterShuffle.

### Why are the changes needed?
While we use the old shuffle fetch protocol, we use partitionId as mapId in the ShuffleBlockId construction,but we use `context.taskAttemptId()` as mapId that it is cached in `taskIdMapsForShuffle` when we `getWriter[K, V]`.

where data could not be cleaned up when unregisterShuffle ,because we remove a shuffle's metadata from the `taskIdMapsForShuffle`'s mapIds, the mapId is `context.taskAttemptId()` instead of partitionId.

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
add new test.

### CODE REVIEW

**Comment by yikf** [comment]
gentle ping @srowen @otterc  @HyukjinKwon  , thanks for taking a look.

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 1]
```diff
@@ -17,13 +17,23 @@
...
 import org.mockito.invocation.InvocationOnMock
 import org.mockito.stubbing.Answer
 import org.scalatest.matchers.must.Matchers
+import org.scalatest.matchers.should.Matchers._
 
 import org.apache.spark._
+import org.apache.spark.rdd.ShuffledRDD
 import org.apache.spark.serializer.{JavaSerializer, KryoSerializer, Serializer}
+import org.apache.spark.util.Utils
+
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:36]: nit: redundant empty line.

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 2]
```diff
@@ -131,4 +141,29 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
     )))
   }
 
+  test("SPARK-34541 Data could not be cleaned up when unregisterShuffle") {
+    val conf = new SparkConf(loadDefaults = false)
+    val tempDir: File = Utils.createTempDir()
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:146]: use `withTempDir`?

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 3]
```diff
@@ -131,4 +141,29 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
...
+    val filesCreatedByShuffle = getAllFiles -- filesBeforeShuffle
+    filesCreatedByShuffle.map(_.getName) should be
+    Set("shuffle_0_0_0.data", "shuffle_0_0_0.index")
+    // Check that the cleanup actually removes the files
+    sc.env.blockManager.master.removeShuffle(0, blocking = true)
+    for (file <- filesCreatedByShuffle) {
+      assert (!file.exists(), s"Shuffle file $file was not cleaned up")
+    }
+  }
+
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:168]: nit: unnecessary line

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 4]
```diff
@@ -131,4 +141,29 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
     )))
   }
 
+  test("SPARK-34541 Data could not be cleaned up when unregisterShuffle") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:144]: nit: `SPARK-34541:`

**Review by Ngone51 — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala (hunk 4 above)
Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala]: Updated.

**Comment by Ngone51** [comment]
cc @xuanyuanking

**Comment by srowen** [comment]
Jenkins retest this please

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 5]
```diff
@@ -131,4 +139,29 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
     )))
   }
 
+  test("Data could not be cleaned up when unregisterShuffle") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: nit: `SPARK-34541: Data ...`

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: BTW, this test requires `spark.shuffle.useOldFetchProtocol=true`?

And it's better to test both `true` and `false`.

Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: It seems that A and B are the same, both increasing from 0. I don't understand why the protocol should be differentiated on the write side

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: We can run a simple job before our target job to make the `taskAttemptId` starts from 1. e.g., 

`sc.parallelize(1 to 10, 1).count()`.

I tried this way and the issue can be reproduced.

Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: OK, thank you. I'll add test later, And I don`t understand why the protocol should be differentiated on the WriteSide, As follow:
`ShuffleMapTask#runTask

   // While we use the old shuffle fetch protocol, we use partitionId as mapId in the
    // ShuffleBlockId construction.
    val mapId = if (SparkEnv.get.conf.get(config.SHUFFLE_USE_OLD_FETCH_PROTOCOL)) {
      partitionId
    } else context.taskAttemptId()
`

Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: I roughly remember that's because we want to ensure the unique file name at write size. cc @xuanyuanking

Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala]: updated

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 6]
```diff
@@ -131,4 +139,40 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
     )))
   }
 
+  test("Shuffle data can be cleaned up whether spark.shuffle.useOldFetchProtocol=true/false") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: We usually test different config values like this:

```scala
Seq(true, false).foreach { value =>
   test("Shuffle data can be cleaned up whether spark.shuffle.useOldFetchProtocol=value") {
     ....
    conf.set(spark.shuffle.useOldFetchProtocol, value)
    ...
}
```

Could you follow this way?

Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala]: ok, Thanks direction for codeStyle very much,xiexie~

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala (hunk 5 above)
Comment (xuanyuanking) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala]: `at write size` -> `at write side`? :)
Yes, you can check the description in #25620. TL;DR: we need a unique file name to resolve the indeterminate shuffle issue.

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 7]
```diff
@@ -131,4 +138,33 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
     )))
   }
 
+  Seq("true", "false").foreach { value =>
+    test(s"SPARK-34541: shuffle can be removed when spark.shuffle.useOldFetchProtocol=$value") {
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: Is this possible to add in `ShuffleSuite`. If so, we don't need to consider the `spark.shuffle.useOldFetchProtocol` since `ShuffleOldFetchProtocolSuite` would cover the old protocol case.
If it's hard to move it to ShuffleSuite. I'm also OK with the current code.

**Review by xuanyuanking — APPROVED** [review]
Thanks for the fix! LGTM

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala (hunk 7 above)
Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:142]: If we move it to `ShuffleSuite`, we should consider  the case of `spark.shuffle.useOldFetchProtoclol = false` ,  and different configurations are also more intuitive to test together.

If we need to move it to ShuffleSuite, I'm alse ok.

**Comment by yikf** [comment]
gentle ping@srowen , Coulde you help me merge the pr to master? thanks

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by yikf** [comment]
> Jenkins retest this please

gentle ping @srowen .

**Review by srowen — COMMENTED** [review]
@Ngone51 OK with you?

File: core/src/test/scala/org/apache/spark/ShuffleSuite.scala [hunk 8]
```diff
@@ -419,6 +424,31 @@ abstract class ShuffleSuite extends SparkFunSuite with Matchers with LocalSparkC
 
     manager.unregisterShuffle(0)
   }
+
+  test(s"SPARK-34541: shuffle can be removed when spark.shuffle.useOldFetchProtocol=true") {
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/ShuffleSuite.scala:428]: nit: `SPARK-34541: shuffle files should be removed normally`

(`spark.shuffle.useOldFetchProtocol=true` is not valid for `ShuffleSuite`.)

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala [hunk 9]
```diff
@@ -130,5 +130,4 @@ class SortShuffleManagerSuite extends SparkFunSuite with Matchers {
       mapSideCombine = true
     )))
   }
-
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:133]: nit: revert the unrelated change.

**Review by Ngone51 — COMMENTED** [review]
@srowen  Thanks for the ping.

Overall, the change still looks good to me. I have left another two minor comments. Let's address them and trigger another round test to pass k8s tests.

File: core/src/test/scala/org/apache/spark/ShuffleSuite.scala (hunk 8 above)
Comment (yikf) [reply core/src/test/scala/org/apache/spark/ShuffleSuite.scala:428]: `ShuffleSuite`  is abstract class, ShuffleOldFetchProtocolSuite would cover the `spark.shuffle.useOldFetchProtocol=true` case.

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala (hunk 9 above)
Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:133]: line-133 is empty line, we add test in  `SortShuffleManagerSuite ` before, after https://github.com/apache/spark/pull/31664#discussion_r585669990, move the test to `ShuffleSuite`, So i change the empty line together.

File: core/src/test/scala/org/apache/spark/ShuffleSuite.scala (hunk 8 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/ShuffleSuite.scala:428]: I mean, the test name is not correct for other extended shuffle suites. For other suites, `spark.shuffle.useOldFetchProtocol` should be `false`, right?

File: core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala (hunk 9 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:133]: I know the reason, but it's still not a necessary change, especially when there's no other changes in the file.

Comment (yikf) [reply core/src/test/scala/org/apache/spark/shuffle/sort/SortShuffleManagerSuite.scala:133]: ok，i will revert it.

**Comment by Ngone51** [comment]
LGTM

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Merged to master
